net/portmapper/portmapper.go

```diff
@@ -0,0 +1,601 @@
+		return
+	}
+	c.localPort = localPort
+	c.invalidateMappingsLocked(true)
+}
+
+func (c *Client) gatewayAndSelfIP() (gw, myIP netaddr.IP, ok bool) {
+	gw, myIP, ok = interfaces.LikelyHomeRouterIP()
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !ok {
+		gw = netaddr.IP{}
+		myIP = netaddr.IP{}
+	}
+	if gw != c.lastGW || myIP != c.lastMyIP || !ok {
+		c.lastMyIP = myIP
+		c.lastGW = gw
+		c.invalidateMappingsLocked(true)
+	}
+	return
+}
+
+func (c *Client) invalidateMappingsLocked(releaseOld bool) {
+	if c.pmpMapping != nil {
+		if releaseOld {
+			c.pmpMapping.release()
+		}
+		c.pmpMapping = nil
+	}
+	c.pmpPubIP = netaddr.IP{}
+	c.pmpPubIPTime = time.Time{}
+	c.pcpSawTime = time.Time{}
+	c.uPnPSawTime = time.Time{}
+}
+
+func (c *Client) sawPMPRecently() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.sawPMPRecentlyLocked()
+}
+
+func (c *Client) sawPMPRecentlyLocked() bool {
+	return !c.pmpPubIP.IsZero() && c.pmpPubIPTime.After(time.Now().Add(-trustServiceStillAvailableDuration))
+}
+
+func (c *Client) sawPCPRecently() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.pcpSawTime.After(time.Now().Add(-trustServiceStillAvailableDuration))
+}
+
+func (c *Client) sawUPnPRecently() bool {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return c.uPnPSawTime.After(time.Now().Add(-trustServiceStillAvailableDuration))
+}
+
+func closeCloserOnContextDone(ctx context.Context, c io.Closer) (stop func()) {
+	// Close uc on ctx being done.
+	ctxDone := ctx.Done()
+	if ctxDone == nil {
+		return func() {}
+	}
+	stopWaitDone := make(chan struct{})
+	go func() {
+		select {
+		case <-stopWaitDone:
+		case <-ctxDone:
+			c.Close()
+		}
+	}()
+	return func() { close(stopWaitDone) }
+}
+
+// NoMappingError is returned by CreateOrGetMapping when no NAT
+// mapping could be returned.
+type NoMappingError struct {
+	Err error
+}
+
+func (nme NoMappingError) Unwrap() error { return nme.Err }
+func (nme NoMappingError) Error() string { return fmt.Sprintf("no NAT mapping available: %v", nme.Err) }
+
+func IsNoMappingError(err error) bool {
+	_, ok := err.(NoMappingError)
+	return ok
+}
+
+var (
+	ErrNoPortMappingServices = errors.New("no port mapping services were found")
+	ErrGatewayDetected       = errors.New("failed to look gateway address")
+)
+
+// CreateOrGetMapping either creates a new mapping or returns a cached
+// valid one.
+//
+// If no mapping is available, the error will be of type
+// NoMappingError; see IsNoMappingError.
+func (c *Client) CreateOrGetMapping(ctx context.Context) (external netaddr.IPPort, err error) {
+	gw, myIP, ok := c.gatewayAndSelfIP()
+	if !ok {
+		return netaddr.IPPort{}, NoMappingError{ErrGatewayDetected}
+	}
+
+	c.mu.Lock()
+	localPort := c.localPort
+	m := &pmpMapping{
+		gw:       gw,
+		internal: netaddr.IPPort{IP: myIP, Port: localPort},
+	}
+
+	prevPort := uint16(0) // or 0 if unknown
+
+	// Do we have an existing mapping that's valid?
+	now := time.Now()
+	if m := c.pmpMapping; m != nil {
+		if now.Before(m.useUntil) {
```

Comment:
We should fudge m.useUntil a bit, e.g. renew 10min before the expiry time the gateway gave us. Otherwise we may end up dropping the mapping for a few seconds/minutes until magicsock re-requests it.

Reply:
It's already halved per the RFC:

```
                                d := time.Duration(pres.MappingValidSeconds) * time.Second
                                d /= 2 // renew in half the time                                                                                                                                              
```

Should be sufficient?

Reply:
Oops, yeah, s'fine.